### PR TITLE
updated JPVerb and Greyhole

### PR DESCRIPTION
**Pull request by LFSaw**
Please review.

Generated cpp with current FAUST toolchain (https://github.com/grame-cncm/faust/commit/06e44a5572eb17dea26c9c37a5c1bfa5f9bdac78). Proposed fix for #133.

Changes are mainly in dsp files and used faust architecture. You should be able to just check out the branch and run `cmake/make/make install`. The cpp files are already generated.

### CODE REVIEW

**Comment by LFSaw** [comment]
@snappizz , I added you as a reviewer mostly because you opened #133 . If you don't want to / don't have time/expertise, please pass it along.

If there will not be anyone interested in reviewing until 8.5.2017, I'll merge the PR myself.

**Comment by snappizz** [comment]
I'll give it a look over the weekend.

**Comment by patrickdupuis** [comment]
I did a quick test using the JPverb doc example. The server shows ~66% with these changes. That's up from 16-28% previously. SC is 3.9dev [f80d6e0] running on Fedora 25.

**Comment by LFSaw** [comment]
The only substantial change I made on the dsp-side is to include the prime number table into the dsp-file itself in idfference to how it was done earlier in jprev.h. The rest of the changes you see in the diff are tab-conversions. Sorry for obfuscating my changes that way, I was not careful enough...

Since the last compile (`.dsp`>>`.cpp`), the FAUST way of library includes changed from single lib includes to one global include-them-all plus the use of namespaces (within the `dsp` file). What happened under the hood I don't know...

**Comment by LFSaw** [comment]
I locally reverted the `jprev.h` change and do not see any difference in CPU usage, i.e. with one instance running, I am at the same `37%` average CPU.

**Comment by LFSaw** [comment]
and with the "old" ".cpp" (from current master b967b66), I see a CPU usage of 9-10% average. So there is clearly something odd here.

**Comment by sletz** [comment]
"I'm sad to say I lost a good amount of confidence in the FAUST project today :(" , this is sad.. 

We did big changes in dsp libraries organization last year, trying to better separate the different parts share more code, fully document them. Even if we did our best... old dsp code was sometimes broken, and had to be adapted. 
New library system is *not* supposed to change the resulting generated cpp code, or in a very narrow way, and is not supposed to change the CPU usage the way it seems to be the case in those precise JPverb cases.
 If also changing between the use of jprev.h of new "waveform" base approach does not change CPU use, then something else is happening. We need to compare the complexity of the generated cpp code to understand if something significant has changed on the library side, or the Faust compiler side.

**Comment by LFSaw** [comment]
Let's not get into ranting/etc. but hope for the best for the future of all involved parties :) 
I am very confident and very happy with both FAUST and the FAUST developers, the language itself is super elegant and the developers are superfast in replying to anything I throw at them... plus I am sure that part of the mess is my disability to use `git submodule` for including the faust sources into sc3-plugins (do we want that? why would this be a bad idea?).

Thanks for everyone helping out to look at it!

**Comment by LFSaw** [comment]
Let's not get into ranting/etc. but hope for the best for the future of all involved parties :) 
I am very confident and very happy with both FAUST and the FAUST developers, the language itself is super elegant and the developers are superfast in replying to anything I throw at them... plus I am sure that part of the mess is my disability to use `git submodule` for including the faust sources into sc3-plugins (do we want that? why would this be a bad idea?).

Thanks for everyone helping out to look at it!

**Comment by LFSaw** [comment]
Thanks to @orlarey and @sletz the issue has been resolved. 

@patrickdupuis and @snappizz, please test if cpu usage is back to "normal" for you with the current version (e61d83f)

**Comment by bagong** [comment]
Unfortunately we get an API-mismatch error in msvc-builds:

`````

  "C:\Users\Rainer\Projects\sc\sc3-plugins\build_VS2015_64\Install.vcxproj" (default target) (1) ->
       "C:\Users\Rainer\Projects\sc\sc3-plugins\build_VS2015_64\ALL_BUILD.vcxproj" (default target) (3) ->
       "C:\Users\Rainer\Projects\sc\sc3-plugins\build_VS2015_64\source\Greyhole.vcxproj" (default target) (32) ->
       (ClCompile target) ->
         C:\Users\Rainer\Projects\sc\sc3-plugins\source\DEINDUGens\GreyholeRaw.cpp(1323): error C2375: 'api_version': redefinition; different linkage [C:\Users\Rainer\Projects\sc\sc3-plugins\build_VS2015_64\source\Greyhole.vcxproj]
         C:\Users\Rainer\Projects\sc\sc3-plugins\source\DEINDUGens\GreyholeRaw.cpp(1326): error C2375: 'load': redefinition; different linkage [C:\Users\Rainer\Projects\sc\sc3-plugins\build_VS2015_64\source\Greyhole.vcxproj]


       "C:\Users\Rainer\Projects\sc\sc3-plugins\build_VS2015_64\Install.vcxproj" (default target) (1) ->
       "C:\Users\Rainer\Projects\sc\sc3-plugins\build_VS2015_64\ALL_BUILD.vcxproj" (default target) (3) ->
       "C:\Users\Rainer\Projects\sc\sc3-plugins\build_VS2015_64\source\JPverb.vcxproj" (default target) (33) ->
         C:\Users\Rainer\Projects\sc\sc3-plugins\source\DEINDUGens\JPverbRaw.cpp(1634): error C2375: 'api_version': redefinition; different linkage [C:\Users\Rainer\Projects\sc\sc3-plugins\build_VS2015_64\source\JPverb.vcxproj]
         C:\Users\Rainer\Projects\sc\sc3-plugins\source\DEINDUGens\JPverbRaw.cpp(1637): error C2375: 'load': redefinition; different linkage [C:\Users\Rainer\Projects\sc\sc3-plugins\build_VS2015_64\source\JPverb.vcxproj]
`````

I remember JPverb being a difficult one when sc3-plugins were made msvc-buildable by @llloret . And actually, I don't think they ever worked in msvc builds.

**Comment by llloret** [comment]
Hi,

@bagong, I think they did work, but I needed to submit a patch to the faust
people, but I would need to check my notes. I am pretty sure I saw that
issue, and fixed it...

I will try to have a quick look, but do not have much time at the moment.
Luis

**Comment by LFSaw** [comment]
this seems to be a good test case for supercollider/faust development :)

File: source/DEINDUGens/faust_src/GreyholeRaw.cpp [hunk 1]
```diff
@@ -49,7 +50,7 @@ using namespace std;
 #if defined(__GNUC__) && __GNUC__ >= 4
 # define FAUST_EXPORT __attribute__((visibility("default")))
 #else
-# define FAUST_EXPORT /* NOP */
+# define FAUST_EXPORT  SC_API_EXPORT
```
Comment (brianlheim) [inline source/DEINDUGens/faust_src/GreyholeRaw.cpp]: this may be what's causing the linkage error reported above

**Comment by brianlheim** [comment]
> Unfortunately we get an API-mismatch error in msvc-builds:

It is a linkage error, not api mismatch.

The two functions named in the error message have their declarations wrapped in `extern "C"` but the definitions are not. See https://github.com/supercollider/sc3-plugins/blob/e46f54d8e6b4ea399bc4f9b9d7aa640923146089/source/DEINDUGens/JPverbRaw.cpp#L1397-L1408 and https://github.com/supercollider/sc3-plugins/blob/e46f54d8e6b4ea399bc4f9b9d7aa640923146089/source/DEINDUGens/JPverbRaw.cpp#L1551.

This commit https://github.com/supercollider/sc3-plugins/commit/d0261812232ad149e2b10ef2ab537c71c10cda7c makes it appear that the `extern "C"` linkage is unnecessary altogether.

**Comment by sletz** [comment]
If needed, please send us (merge request for Faust Github..) changes in the original https://github.com/grame-cncm/faust/blob/master-dev/architecture/supercollider.cpp file

**Comment by LFSaw** [comment]
> If needed, please send us (merge request for Faust Github..) changes in the original https://github.com/grame-cncm/faust/blob/master-dev/architecture/supercollider.cpp file

I think it's needed but since I'm not on windows, I cannot try/test this... 
@llloret did you submit the PR back then? Could you do now? I can then update this PR with said changes.

**Comment by llloret** [comment]
Yes, I think I submitted a PR and was taken then. I would need to check to
make sure, but can't at the moment...

**Comment by LFSaw** [comment]
@llloret , weird that it is not in the current setting then...

**Comment by llloret** [comment]
Ok,

it wasn't done through a formal PR, but through email: these are a couple
of mails I sent:

+++++++++++++++++++++++++++++++++++ MAILS
 +++++++++++++++++++++++++++++++++++++

Hi,

There has been some effort to port the SuperCollider external plugins
repository to MSVC 64 bits, and that required some changes to the
Faust generated sources.

I was wondering if someone could make those changes available
upstream, so that we can generate the sources automatically in the
future (or alternatively, point me to where I could make that myself,
and make it available upstream).

I have seen in previous messages in the list that Alfonso Santimone,
which I have also seen around in SC, was planning to make something
around these lines, and I was wondering if this is something that
could be done as part of that effort?

The changes are minimal, basically:

- Change "# define FAUST_EXPORT /* NOP */"
to "# define FAUST_EXPORT SC_API_EXPORT"

- Change "int api_version(void);"
to "FAUST_EXPORT int api_version(void);"

- Change "void load(InterfaceTable*);"
to FAUST_EXPORT void load(InterfaceTable*);

You can also look at the changes in the PR I issued for the
Supercollider plugins here:
https://github.com/supercollider/sc3-plugins/pull/80/files, and then
looking for "FAUST".

As I said this changes would only be necessary for SuperCollider
generation under Windows.

Thanks,
Luis

Hi, Yann,

I have seen that you took part of the changes that I asked for to
allow for Compilation with MSVC under Windows. Thanks for that.

But from what I see, only part of the changes where done. I can see
the change of the #define, but I cannot see the addition of
FAUST_EXPORT in front of "api_version" and "load".

Could you please have a look?
Thanks again,
Luis

+++++++++++++++++++++++++++++++++++++++++ END OF MAILS
+++++++++++++++++++++++++++++

So, it seems like my suggestions were not fully taken... take into account
this was about one year ago, so I might be missing something... In
particular, perhaps Yann replied, but I cannot find that email. Anyway, it
should be easy to fix it, I think.

Luis

**Comment by brianlheim** [comment]
See this PR - https://github.com/supercollider/sc3-plugins/pull/80

All that needs to be done is make sure the linkages match for those two functions. Hard for me to tell at a glance if there has been a consistent way of doing that, but it seems like the declarations should probably not be given C linkage.

**Comment by sletz** [comment]
OK, since Faust is now (since recently.. on github) can you do a clean PR then? Thanks.

**Comment by LFSaw** [comment]
just for future reference, please do not manually change/"fix" FAUST-generated .cpp-files (as JPVerbRaw.cpp in DEINDUGens) since those are, well, generated...

**Comment by LFSaw** [comment]
on a side note, @snappizz requested changes are now successfully handled? How do I get rid of the "Changes requested"-flag? Do I "dismiss review"?

**Comment by brianlheim** [comment]
@sletz I am not very involved with this repo, have actually never built it. I'd prefer that somebody else do it.

**Comment by sletz** [comment]
"just for future reference, please do not manually change/"fix" FAUST-generated .cpp-files (as JPVerbRaw.cpp in DEINDUGens) since those are, well, generated..." 

Yes this is the point, correcting the Faust architecture file in the first place...

**Comment by brianlheim** [comment]
> on a side note, @snappizz requested changes are now successfully handled? How do I get rid of the "Changes requested"-flag? Do I "dismiss review"?

"Changes requested" goes away when the original reviewer submits an "Approved" review. Someone else can also do it, but it'll look like 1 X / 1 ✓

**Comment by llloret** [comment]
I'll do the PR on faust, and see if it fixes the issue.

**Comment by llloret** [comment]
@sletz, I have just created a PR. It's number 34 on faust github.

**Comment by llloret** [comment]
@sletz, I have done the PR against the faust2 branch, since it seemed the right place... should I do it somewhere else?

**Comment by bagong** [comment]
Good news: with @llloret changes, both plugins build and work. Work, as in the first help example works. The second example contains a dead link. I guess the succeeding code depends on that.

I do seem to remember that there was an example with JPverb on sc-users that didn't work with the VS-build (but worked with the mingw-build)? @snappizz , do you remember?

**Comment by LFSaw** [comment]
yes, to master-dev :)

**Comment by LFSaw** [comment]
Thanks for spotting the dead links, @bagong, I fixed them.

**Comment by sletz** [comment]
"@sletz, I have done the PR against the faust2 branch, since it seemed the right place... should I do it somewhere else?" 

Nope, please do the PR for master-dev branch.

**Comment by llloret** [comment]
Ok, will do later today

>

**Comment by bagong** [comment]
Found the example that used to crash the msvc-build:

http://new-supercollider-mailing-lists-forums-use-these.2681727.n2.nabble.com/JPVerb-broken-td7631024.html#a7631043

````
(
SynthDef(\bpfsaw, {
    |atk=2, sus=0, rel=3, c1=1, c2=(-1), freq=500, detune=0.1, cf=1500, rq=0.2, amp=1, out=0|
    var sig, env;
    env = EnvGen.kr(Env([0,1,1,0],[atk,sus,rel],[c1,0,c2]),doneAction:2);
    sig = Saw.ar(freq * LFNoise1.kr(0.1,detune).midiratio);
    sig = BPF.ar(sig, cf, rq);
    sig = sig * env * amp;
    //sig = JPverb.ar(sig);
    Out.ar(out, sig);
}).add;
)

x = Synth(\bpfsaw);
````

No more crash ;) Thanks!

**Comment by llloret** [comment]
Hi, @sletz, I have now submitted a new PR with the required change against the master-dev branch as you requested. It is PR 35.

**Comment by sletz** [comment]
PR done.

**Comment by llloret** [comment]
Brilliant, thanks!

**Comment by LFSaw** [comment]
(Dear future readers, this is the pull request: https://github.com/grame-cncm/faust/pull/35)

**Comment by LFSaw** [comment]
updated PR, please test on msvc ( @bagong ?)

**Comment by LFSaw** [comment]
@snappizz , could you please review your change request and mark it resolved, if you feel it is?

**Comment by patrickdupuis** [comment]
Tested this again. I'm still seeing an increase, although it is down to ~40%.

**Comment by LFSaw** [comment]
@patrickdupuis , would you mind sharing your test environment and build settings, please? I don;t see these increases.

**Comment by patrickdupuis** [comment]
Sure! Let me know if you need anything additional.
Linux 4.10.14, Fedora 25 64-bit, GCC 6.3.1
SC 3.9dev [f80d6e0]
both SC and plugins are built NATIVE=ON, otherwise default cmake config.

**Comment by LFSaw** [comment]
I am on OSX, so cannot test on linux... could someone else please confirm the increase on the linux platform?

Just to be clear, we are now talking about comparing the current `sc3-plugins` head with this PR. Also, make sure to run `make install` (I missed this several times in this endeavour...) before testing.

Test code would be:

```
(
// there is no need to re-open the IDE, just reboot the server after recompiling plugins
s.options.memSize_(65536 * 4);
s.reboot;
)

(
Ndef(\hole, {
    var src, delay, amp;

    src = Splay.ar(Impulse.ar([1, 3, 5, 7, 9]).scramble);
    delay = Greyhole.ar(
        src,
        \dtime.kr(0.1),
        \damp.kr(0.1),
        \size.kr(1),
        \diff.kr(0.707),
        \fb.kr(0.1),
        \modDepth.kr(0.01),
        \modFreq.kr(2)
    );
    delay
});
)
```

**Comment by patrickdupuis** [comment]
Running the above example:
(the server idles at ~2.5%)
sc3plugins master => 18-25% (fluctuates a lot)
this PR => ~37% (not much fluctuation)

**Comment by LFSaw** [comment]
anyone else has an idea? @sletz, do the changes you introduced in https://github.com/grame-cncm/faust/commit/23413d6ce18a8b408ed512270b4a179252e87833 also affect the linux compile, or is it OSX-only?

**Comment by sletz** [comment]
This commit changes the generated cpp code, so it should behave absolutely the same on OS X and Linux.

**Comment by LFSaw** [comment]
> This commit changes the generated cpp code, so it should behave absolutely the same on OS X and Linux.

thanks for the clarification, @sletz, I was suspecting this. Still, I do not see the reported increase of cpu usage on my (OSX 10.12.4 (16E195)) machine... could there be other things that changed since faust 0.9.62 regarding e.g. compiler optimisation/llvm/etc? 

IMHO, an increase of factor 1.5 in cpu usage is quite significant and it would be great to find the cause of this.

**Comment by LFSaw** [comment]
compile flags used in `faust2supercollider`: 

```
CXXFLAGS = -O3 -xT -ftz -fno-alias -fp-model fast=2
```

compile flags I found for my cmake-generated file (on osx):

```
CXX_FLAGS =  -msse -msse2 -mfpmath=sse -std=c++11 -stdlib=libc++ -fPIC   -fvisibility=hidden
```

@sletz , could you have a look, please? I found those in `build/source/CMakeFiles/Greyhole.dir/flags.make`

**Comment by sletz** [comment]
What is this " cmake-generated file (on OS X):" ?
cpp code should be compiled with same flags for proper comparaison obviously...

**Comment by LFSaw** [comment]
The `cmake`-generated file is what the `CMakeLists.txt` generates for the automated build environment in sc3-plugins. 
The checks I run on my OSX machine were build with this cmake-generated flags (both master head and this PR), and I did not experience the increase in cpu-resource-hungry-ness. 
My build environment:

`Apple LLVM version 8.1.0 (clang-802.0.42)`

So, I was wondering if the clang compiler implicitly assumes optimisations, whereas the linux gcc-toolchain might not do this.

**Comment by bagong** [comment]
I checked on Windows, Linux and Mac now and cannot confirm the high cpu usage observed. This is with your example from above, right @LFSaw , with .play added.

- Windows-msvc build is fine now, the changes in faust generated the sc-code as expected
- on a Lenovo X201 running Win10 SCWin 3.9dev 64-bit msvc-build your example eats 9%
- on a Lenovo X200 running Ubuntu Studio 17.04 default 3.9 build (release build!) it eats 10%
- on a MacBook MacOS Sierra 2012 64-bit release it eats somewhere between 3 and 9%.

First and second percentage-frames mean average and top useage, right? Anyways, on OSX they are pretty far apart (3/9%), on Windows an Linux they are similar and constant in the 10% region. Note that the Linux and Windows machines are far weaker computers than the MacBook.

@LFSaw , you seem to have forgotten to add the updated links in help. And actually, how about updating the second examples in a way that they don't produce all those warnings? They don't have a meaning, and yet...

Not sure about that optimizations question. I might be saying something stupid now, but be it so: in this situation Faust only generates code, right, it doesn't actually create a binary? I'd be surprised if compiler optimizations would influence the generated code.

**Comment by LFSaw** [comment]
Thanks for looking into this, @bagong . 
It is strange that @patrickdupuis still sees so major differences in his compiled versions...

Help-file links: 
5698bd3 adds those (or do I miss something?)

Warnings: 
they appear because the two used UGens have slightly different argument interfaces. Since I choose randomly between the two, I cannot foresee, which one is used… I am currently not aware of an alternative writing of the mentioned code, without making it unnecessarily complicated.
I therefore added a note on the warnings to the help files.

**Comment by bagong** [comment]
Ah, sorry, probably my build must have still been using the help files that had already been generated...

**Comment by patrickdupuis** [comment]
I have tested again with SC and sc3plugins built NATIVE=OFF and CMAKE_BUILD_TYPE=Release. The performance is now ~10% as reported by @bagong.

**Comment by patrickdupuis** [comment]
Oddly, the real performance hit seemed to be due to SC being build NATIVE=ON.

**Comment by LFSaw** [comment]
@snappizz I think we're good to go here, could you please mark your change request as "resolved"?

anyone, please merge to master.

**Comment by snappizz** [comment]
@LFSaw First I'll need to take a look at it. Hopefully I'll have time for that tomorrow.

**Comment by patrickdupuis** [comment]
The beginning of the examples section of Geryhole.schelp doesn't look right. Something is missing in order to make the code appear as it should.

**Comment by LFSaw** [comment]
fixed.

**Comment by bagong** [comment]
@patrickdupuis , I was curious in things "native" (a nice tweaker toy ;)) and checked on the Ubuntu machine: if anything, it get's a bit faster for me, I am seeing something around 8% stable now. I still suspect that you accidentally undefining CMAKE_BUILD_TYPE is at the core of your unusual observations. Note that build settings you once made on the command line stick over builds until you either:

- delete the file CMakeCache.txt in your build folder and reconfigure the build
- build from a empty new build folder
- reassign that particular variable with a correctly spelt alternative on the command line ;) So you would have had to do something like `cmake -D CMAKE_BUILD_TYPE=Release` (or RelWithDebInfo) to get rid of that misconfiguration. Just rerunning `cmake ..` will not reset variables assignments made on the command line to default values, even if they were wrong ;)

Btw, I am still curious about some insight into the question how FAUST compiler flags would influence code generation and interact with sc3-plugins compiler flags on building. I find FAUST a very cool system we should be interacting with well. But maybe that is outside of the scope of this thread. Then just ignore it. Thanks.

**Comment by sletz** [comment]
Faust compiler flags surely influence the c++ code generation and possibly allow to generate code that is faster/slower depending of the source DSP code.

We developed tools that allow to automatically "bench" a given DSP for several Faust compiler flags. It may be useful to use it to optimize Faust sc3-plugins, see the post here:
http://faust.grame.fr/news/2017/04/26/optimizing-compilation-parameters.html

**Comment by LFSaw** [comment]
@sletz – cool! 

I tried running it (naïvely) with current faust/master-dev https://github.com/grame-cncm/faust/commit/c4af503b17e016e1364ad1adbe26adb20a9ed67c

```
$ faustbench JPverbRaw.dsp
```
which resulted in
```
Selected compiler is g++ with CXXFLAGS = -Ofast
/usr/local/share/faust/faustbench.cpp:34:10: fatal error: 'dsp_scal.h' file not found
#include "dsp_scal.h"
         ^
1 error generated.
/usr/local/bin/faustbench: line 120: ./JPverbRaw: No such file or directory
```

Also, is there a possibility (and need) to include the `jprev.h` file somehow?

----

Although cool, I think this is outside the scope of this PR.

**Comment by sletz** [comment]
You've tested on Linux right? (not much tested I have to say ((-;)
Can you try this fix : https://github.com/grame-cncm/faust/commit/1d302fb797c648af0947b41273c68de7675e1153

**Comment by LFSaw** [comment]
nope, osx... tried anyway (got double-header definition error for `jprev.h`, thanks for "spotting" :) ) and it seems to work, benching successful!

**Comment by LFSaw** [comment]
... and _wow_, this dropped performance on my machine by more than 50% (i.e. average of 3% CPU usage instead of previous 9%) 

cool!

@bagong , would you mind testing the updated PR on your multi-platform machine park?

**Comment by sletz** [comment]
Can you paste the result of "faustbench JPverbRaw.dsp" ? That would be interesting !

**Comment by LFSaw** [comment]
```sh
$ faustbench JPverbRaw.dsp
Selected compiler is g++ with CXXFLAGS = -Ofast-fbracket-depth=512
DSP bench of ./JPverbRaw compiled in C++
-scal : 30.6195
-vec -lv 0 -vs 4 : 21.1588
-vec -lv 0 -vs 8 : 38.1386
-vec -lv 0 -vs 16 : 36.4325
-vec -lv 0 -vs 32 : 33.6898
-vec -lv 0 -vs 64 : 32.603
-vec -lv 0 -vs 128 : 31.0563
-vec -lv 0 -vs 256 : 29.5692
-vec -lv 0 -vs 512 : 27.151
-vec -lv 0 -vs 4 -g : 21.1834
-vec -lv 0 -vs 8 -g : 37.8544
-vec -lv 0 -vs 16 -g : 36.4108
-vec -lv 0 -vs 32 -g : 33.0877
-vec -lv 0 -vs 64 -g : 32.298
-vec -lv 0 -vs 128 -g : 31.4482
-vec -lv 0 -vs 256 -g : 29.5999
-vec -lv 0 -vs 512 -g : 27.4923
-vec -lv 1 -vs 4 : 21.16
-vec -lv 1 -vs 8 : 27.7116
-vec -lv 1 -vs 16 : 31.424
-vec -lv 1 -vs 32 : 31.9556
-vec -lv 1 -vs 64 : 31.6609
-vec -lv 1 -vs 128 : 31.1959
-vec -lv 1 -vs 256 : 29.068
-vec -lv 1 -vs 512 : 27.221
-vec -lv 1 -vs 4 -g : 20.5372
-vec -lv 1 -vs 8 -g : 27.5882
-vec -lv 1 -vs 16 -g : 31.2704
-vec -lv 1 -vs 32 -g : 31.8408
-vec -lv 1 -vs 64 -g : 31.7942
-vec -lv 1 -vs 128 -g : 31.0613
-vec -lv 1 -vs 256 -g : 29.2739
-vec -lv 1 -vs 512 -g : 27.3036
Best value is : 38.1386 with -vec -lv 0 -vs 8
```

**Comment by bagong** [comment]
I love the direction this thread is taking!

My results are interesting and kind of a mixed bag. On Ubuntu 17.04 (gcc 6.3) build as usual (I retained the native=on setting, can switch it off if that's better) the example got *more hungry*:  17% (I rebooted to verify, it really went down that far)

On Windows msvc VS2015 something interesting happened: the , the build got stuck for about 15 minutes (I thought it had died), and then it actually continued and completed successfully. The stop occured here:

````
C:\Users\Rainer\Projects\sc\sc3-plugins\source\DEINDUGens\GreyholeRaw.cpp(4309): warning C4200: nonstandard extension used: zero-sized array in struct/union [C:\Users\Rainer\Projects\sc\sc3-plugins\build_VS2015_64\source\Greyhole.vcxproj]
  C:\Users\Rainer\Projects\sc\sc3-plugins\source\DEINDUGens\GreyholeRaw.cpp(4309): note: This member will be ignored by a defaulted constructor or copy/move assignment operator
C:\Users\Rainer\Projects\sc\sc3-plugins\source\DEINDUGens\GreyholeRaw.cpp(4333): warning C4996: 'strdup': The POSIX name for this item is deprecated. Instead, use the ISO C and C++ conformant name: _strdup. See online help for details. [C:\Users\Rainer\Projects\sc\sc3-plugins\build_VS2015_64\source\Greyhole.vcxproj]
  C:\Program Files (x86)\Windows Kits\10\Include\10.0.15063.0\ucrt\string.h(538): note: see declaration of 'strdup'
C:\Users\Rainer\Projects\sc\sc3-plugins\source\DEINDUGens\GreyholeRaw.cpp(4416): warning C4267: 'initializing': conversion from 'size_t' to 'int', possible loss of data [C:\Users\Rainer\Projects\sc\sc3-plugins\build_VS2015_64\source\Greyhole.vcxproj]
````

Processor consumption is in the 12% area, so it went down a bit too.
The example that used to crash the server in msvc build still works.

So for the MSVC build I'd say little change in performance, but kind of a weird and not very fortunate behavious in the build. Most people will probably think that the build broke.

Mac I cannot test atm. Something is wrong with my network and I can only reboot in an hour or so. Will let you know then.

Maybe `faustbench`optimizes for the local system? Or is a bit Apple centric?

Best, and thanks for the efforts!

**Comment by LFSaw** [comment]
AFAIK `faustbench` does not optimise but tests several build options against each other (see previous listing), so it generates and compiles different versions of the cpp-code and evaluates them against each other. On my local machine, of course...

So, question is: How do we deal with this? 

1. leave the cpp generated by default faust settings, explaining in the README how to do custom optimised builds
2. run `faustbench` on a bunch of different machines, settling for the average optimisation
3. persuade FAUST developers to create a cross-platform optimisation directive that generates a `cpp` file that recognises the current platform and optimises accordingly. 

Although I like *3*, it will not happen and will also introduce new obstacles: what about pre-compiled downloads?
*2* seems a little too ambitious to me, which leaves us with a reasonable *1*.

**Comment by bagong** [comment]
I agree, 1 is fine! Cross-platform optimization efforts often introduce other problems. The SC build is also generic, if people want to tweak, they have to build themselves or ask a friend ;)

But documentation, e.g. in the SC Wiki, that also keeps in mind that not everybody is a build-shark would be super ace!

**Comment by bagong** [comment]
Btw, the "NATIVE" compile-flag seems to make no difference for the deind-ugens (and for most others) in that Ubuntu environment. The binaries are exactly the same.

**Comment by sletz** [comment]
2) "run faustbench on a bunch of different machines, settling for the average optimisation" ==> this does not make much sense, since  the final result of **faustbench** also depends of version of the used c++ compiler..etc.. The point is to find the best Faust compilation parameters for a given hardware/os/compiler 3) is too ambitious for now.. so 1) is the  reasonable way...

**Comment by LFSaw** [comment]
Documentation on how to build is here: 

https://github.com/LFSaw/sc3-plugins/blob/topic/JPVerbGreyholeFix/source/DEINDUGens/faust_src/README.md

**Comment by sletz** [comment]
https://github.com/grame-cncm/faust/commit/35b46311d8e988832968d0651239d15175cae545
should remove compilation warning on Windows.

**Comment by LFSaw** [comment]
I reverted the optimisation and instead documented how to perform the benchmark tests in https://github.com/LFSaw/sc3-plugins/blob/topic/JPVerbGreyholeFix/source/DEINDUGens/faust_src/README.md

**Comment by sletz** [comment]
Anyone interested in working on embedding the Faust compiler directly in SuperCollider ? That would be even better... See this post :
http://faust.grame.fr/news/2016/01/12/using-dynamic-compiler.html

**Comment by LFSaw** [comment]
A JIT-style faust integration into SuperCollider would be insanely great to have. However, there are certain caveats for this built right into the SC architecture:

SuperCollider's architecture is modular: 

image

while the language deals with high-level control structures and is, as an interpreted language, highly flexible, sound processing happens on `scsynth`, a hardwired structure of a node tree, audio/control buses and `(float*)` buffers.

A trivial choice to put the JIT compiler would be an scsynth plugin. However, operation of any scsynth plugin requires it to be realtime-save, which in particular means that the `CTor` cannot take too long. >> bummer.

Another choice would be to integrate JIT compiler into the language. Apart from that C++ extensions of the language are far from non-trivial (and fellow developers will properly laugh at us trying to get yet another feature implemented there), this would mean that the compiled program needs to be transferred to the server somehow... almost the only possibility for data exchange between language and server are OSC messages, the only data type available would be buffer objects that can only contain floats (well, we could cast those into whatever... but that's kinda ugly).

But maybe this is something for a different place to discuss... maybe [sc-dev](http://new-supercollider-mailing-lists-forums-use-these.2681727.n2.nabble.com/)?

**Comment by sletz** [comment]
OK I see..
"But maybe this is something for a different place to discuss... maybe sc-dev?" ==> then I suggest you start the discussion there...

**Comment by LFSaw** [comment]
http://new-supercollider-mailing-lists-forums-use-these.2681727.n2.nabble.com/LLVM-based-libfaust-integration-td7632284.html

**Comment by LFSaw** [comment]
Not wanting to press anything but I would be more than happy to merge this PR so that we (I) can move on to other things. All issues with it seem to be fixed and discussions are diverting into only loosely related topics. Please @snappizz, could you review your change request or pass it on to someone else?

**Comment by snappizz** [comment]
On my system, this PR ups the CPU usage from around 9-10% to 11-13%. I guess that's alright.

Please have patience... I would like to look over the code before approving it.

**Comment by patrickdupuis** [comment]
@snappizz try testing with your CPU governor set to `performance`. It brings it down from 11-13% to 5-6% on my machine, which is an improvement from before.

**Comment by LFSaw** [comment]
again, @snappizz , please review or release for others to have a look. I am sorry to say that your passiveness is getting increasingly annoying. I do understand that you might have other and more important issues to do (and I appreciate every minute you put into maintaining SC3-related code) but honestly, here you are just dragging things out by not releasing yourself from the self-imposed duty to review changes you requested very vaguely earlier on.

I (and several others) took time and effort to fix issues with these UGens which are (a) in sc-plugins, not in the main distro and therefore not meant to be as thoroughly reviewed  and (b) made available by me and my colleague Julian Parker back in the days, so I claim (co-)authorship to them. I thought I increase its quality not only by investing time and effort towards making them more stable on supported platforms but also by opening up its reviewing process. It helped a lot but this passiveness makes me shaky and I increasingly understand @redFrik's decision to revoke his pull request in #141 . Please please please, either give constructive feedback, or mark your request as resolved (by reviewing the code or by giving up your reviewer status).

**Comment by LFSaw** [comment]
thanks everyone involved, it was a pleasure :)

**Comment by vivid-synth** [comment]
> almost the only possibility for data exchange between language and server are OSC messages, the only data type available would be buffer objects that can only contain floats (well, we could cast those into whatever... but that's kinda ugly).

OSC messages can contain binary blobs - for example, that's how we send synthdefs.

**Pull request merged**